dev/src/index.ts

```diff
@@ -1198,15 +1219,96 @@ export class Firestore implements firestore.Firestore {
     this.bulkWritersCount -= 1;
   }
 
+  /**
+   * Recursively deletes all documents and subcollections at and under the
+   * specified level.
+   *
+   * If any deletes fail, the promise is rejected with an error message
+   * containing the number of failed deletes and the stack trace of the last
+   * failed delete. The provided reference is deleted regardless of whether
+   * all deletes succeeded, except when Firestore fails to fetch the provided
+   * reference's descendants.
+   *
+   * Firestore uses a BulkWriter instance with default settings to perform the
+   * deletes. To customize throttling rates or add success/error callbacks,
+   * pass in a custom BulkWriter instance.
+   *
+   * @param ref The reference of a document or collection to delete.
+   * @param bulkWriter Custom BulkWriter instance used to perform the deletes.
+   * @return A promise that resolves when all deletes have been performed.
+   * The promise is rejected if any of the deletes fail.
+   */
+  recursiveDelete<T = DocumentData>(
+    ref: CollectionReference<T> | DocumentReference<T>,
+    bulkWriter?: BulkWriter
+  ): Promise<void> {
+    const writer = bulkWriter ?? this.getBulkWriter();
+    writer._verifyNotClosed();
+    const docStream = this.getAllDescendants(ref);
+    const deleteCompletedDeferred = new Deferred<void>();
+    let errorCount = 0;
+    let lastError: Error | undefined;
+    const incrementErrorCount = (err: Error): void => {
+      errorCount++;
+      lastError = err;
+    };
+    const onStreamEnd = (): void => {
+      writer.flush().then(async () => {
+        if (ref instanceof DocumentReference) {
+          writer.delete(ref).catch(err => incrementErrorCount(err));
+          await writer.flush();
+        }
+
+        if (lastError === undefined) {
+          deleteCompletedDeferred.resolve();
+        } else {
+          let error = new GoogleError(
+            `${errorCount} ` +
+              `${errorCount > 1 ? 'deletes' : 'delete'} ` +
+              'failed. The last delete failed with: '
+          );
+          if (lastError instanceof BulkWriterError) {
+            error.code = (lastError.code as number) as Status;
+          }
+          error = wrapError(error, stack);
+
+          // Wrap the BulkWriter error last to provide the full stack trace.
+          deleteCompletedDeferred.reject(
+            lastError.stack ? wrapError(error, lastError.stack ?? '') : error
+          );
+        }
+      });
+    };
+
+    // Capture the error stack to preserve stack tracing across async calls.
+    const stack = Error().stack!;
+
+    docStream
+      .on('error', err => {
+        err.code = Status.FAILED_PRECONDITION;
+        err.message =
+          'Failed to fetch children documents. ' +
+          'The provided reference was not deleted.';
+        incrementErrorCount(err);
```

Comment:
This shouldn't bump up the error count that is used for the error message, should it?